OPAL/apk/src/main/scala/org/opalj/apk/analyses/ContextRegisteredReceiversAnalysis.scala

```diff
@@ -0,0 +1,144 @@
+                                classHierarchyMatches(project, call.declaringClass.mostPreciseObjectType)) {
+                                val receiverType = call.params.head.asVar.value.asReferenceValue.upperTypeBound
+                                // check if broadcast receiver param is null
+                                // if yes: sticky intent, result is cached, no code executed -> ignore
+                                if (receiverType.nonEmpty) {
+                                    // get broadcast receiver class, might be imprecise
+                                    val receiverClass = receiverType.head.toJava
+
+                                    // try to find intents, might be incomplete
+                                    val intentDef = tacMethod.stmts(call.params(1).asVar.definedBy.head)
+                                    val (actions, categories) = assembleIntentFilter(tacMethod, intentDef)
+
+                                    foundReceivers.append(new ApkContextRegisteredReceiver(
+                                        receiverClass, actions, categories, m, s.pc
+                                    ))
+                                }
+                            }
+                        })
+                    }
+                case _ =>
+            }
+        }))
+
+        // TODO calls from native code
+
+        foundReceivers.toSeq
+    }
+
+    private def classMatches(clazz: ObjectType): Boolean = {
+        clazz.toJava.equals(ContextClass) || clazz.toJava.equals(LocalBroadcastManagerClass) ||
+            clazz.toJava.equals(ActivityClass)
+    }
+
+    private def classHierarchyMatches(project: Project[_], clazz: ObjectType): Boolean = {
+        var tmpClazz = clazz
+        while (!classMatches(tmpClazz)) {
+            if (tmpClazz.toJava.equals("java.lang.Object")) {
+                return false
+            }
+            tmpClazz = project.classFile(tmpClazz) match {
+                case Some(c) => c.superclassType.get
+                case _       => return false
+            }
+        }
+        true
+    }
+
+    /**
+     * Tries to rebuild the IntentFilter for a registerReceiver() call. Only works with IntentFilters that are
+     * created and its actions and categories are added in the same method as where registerReceiver() is called.
+     */
+    private def assembleIntentFilter(
+        tacMethod: AITACode[TACMethodParameter, ValueInformation],
+        intentDef: Stmt[DUVar[ValueInformation]]
+    ): (Seq[String], Seq[String]) = {
+        val foundActions: ListBuffer[String] = ListBuffer.empty
+        val foundCategories: ListBuffer[String] = ListBuffer.empty
+        if (intentDef.isAssignment && intentDef.asAssignment.expr.isNew &&
+            intentDef.asAssignment.expr.asNew.tpe.toJava.equals(IntentFilterClass)) {
+            intentDef.asAssignment.targetVar.usedBy.map(tacMethod.stmts(_)).foreach(s => {
+                var isAction = true
```

Comment:
Don't use a var, the conditions can just return a tuple `(VirtualMethodCall, Boolean)`